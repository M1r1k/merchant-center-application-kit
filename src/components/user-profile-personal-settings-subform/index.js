export { default } from './user-profile-personal-settings-subform';
