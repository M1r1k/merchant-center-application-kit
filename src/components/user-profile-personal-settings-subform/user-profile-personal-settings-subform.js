import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { compose } from 'recompose';
import { defaultMemoize } from 'reselect';
import Select from 'react-select';
import {
  withTimeZones,
  timeZonesShape,
} from '@commercetools-local/l10n/time-zone-information';
import CollapsiblePanel from '@commercetools-local/ui-kit/panels/collapsible-panel';
import FormBox from '@commercetools-local/core/components/form-box';
import LabelField from '@commercetools-local/core/components/fields/label-field';
import { withUser } from '../fetch-user';
import messages from './messages';

export const timeZonesToOptions = defaultMemoize(timeZones =>
  Object.entries(timeZones).map(([code, value]) => ({
    value: code,
    // E.g. `Europe/Berlin - CEST (+02:00)`
    label: `${code} - ${value.abbr} (${value.offset})`,
  }))
);

export const UserProfilePersonalSettingsSubform = props => (
  <CollapsiblePanel label={<FormattedMessage {...messages.title} />}>
    <FormBox>
      <LabelField
        title={<FormattedMessage {...messages.language} />}
        isRequired={true}
      />
      <Select
        name="language"
        value={props.formik.values.language}
        onChange={option => {
          props.formik.setFieldValue('language', option.value);
          props.formik.setFieldTouched('language');
        }}
        onBlur={() => {
          props.formik.setFieldTouched('language');
        }}
        options={[{ value: 'en', label: 'EN' }, { value: 'de', label: 'DE' }]}
        clearable={false}
        searchable={false}
        parse={
          /* transform react select option to form value */
          ({ value }) => value
        }
        disabled={props.formik.isSubmitting}
      />
    </FormBox>
    <FormBox>
      <LabelField title={<FormattedMessage {...messages.timeZone} />} />
      <Select
        name="timeZone"
        value={props.formik.values.timeZone}
        onChange={option => {
          props.formik.setFieldValue('timeZone', option ? option.value : null);
          props.formik.setFieldTouched('timeZone');
        }}
        onBlur={() => {
          props.formik.setFieldTouched('timeZone');
        }}
        options={timeZonesToOptions(props.timeZones)}
        clearable={true}
        searchable={true}
        parse={
          /* transform react select option to form value */
          option => (option ? option.value : null)
        }
        disabled={props.formik.isSubmitting}
      />
    </FormBox>
  </CollapsiblePanel>
);
UserProfilePersonalSettingsSubform.displayName =
  'UserProfilePersonalSettingsSubform';
UserProfilePersonalSettingsSubform.propTypes = {
  formik: PropTypes.shape({
    isSubmitting: PropTypes.bool.isRequired,
    values: PropTypes.shape({
      language: PropTypes.string.isRequired,
      timeZone: PropTypes.string,
    }),
    setFieldValue: PropTypes.func.isRequired,
    setFieldTouched: PropTypes.func.isRequired,
  }),

  // HoC
  timeZones: timeZonesShape,
  user: PropTypes.shape({
    locale: PropTypes.string.isRequired,
  }).isRequired,
};

export default compose(
  withUser(userData => ({
    user: { locale: userData.user && userData.user.language },
  })),
  withTimeZones(ownProps => ownProps.user.locale)
)(UserProfilePersonalSettingsSubform);
