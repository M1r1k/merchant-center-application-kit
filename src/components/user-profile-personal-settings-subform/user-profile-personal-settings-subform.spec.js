import React from 'react';
import { shallow } from 'enzyme';
import { intlMock } from '@commercetools-local/test-utils';
import { UserProfilePersonalSettingsSubform } from './user-profile-personal-settings-subform';

const createFormProps = props => ({
  isSubmitting: false,
  values: {
    language: 'en',
    timeZone: 'Europe/Berlin',
  },
  intl: intlMock,
  setFieldValue: jest.fn(),
  setFieldTouched: jest.fn(),

  ...props,
});
const createTestProps = props => ({
  form: createFormProps(),

  intl: intlMock,
  timeZones: {
    'Europe/Berlin': { name: 'Europe/Berlin', abbr: 'CEST', offset: '+02:00' },
  },
  user: {
    locale: 'en',
  },

  ...props,
});

describe('rendering', () => {
  describe('when not submitting', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<UserProfilePersonalSettingsSubform {...props} />);
    });
    it('should ensure layout structure', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render select for language', () => {
      expect(wrapper).toRender({ name: 'language' });
    });
    it('should render select for timeZone', () => {
      expect(wrapper).toRender({ name: 'timeZone' });
    });
  });

  describe('when submitting', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({
        form: createFormProps({
          isSubmitting: true,
          touched: { firstName: true },
        }),
      });
      wrapper = shallow(<UserProfilePersonalSettingsSubform {...props} />);
    });
    it('should disable all input fields', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should disable the language dropdown', () => {
      expect(wrapper.find({ name: 'language' })).toHaveProp('disabled', true);
    });
  });
});
