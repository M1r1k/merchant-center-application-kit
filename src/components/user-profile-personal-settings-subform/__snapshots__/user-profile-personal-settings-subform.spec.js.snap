// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering when not submitting should ensure layout structure 1`] = `
<CollapsiblePanel
  label={
    <FormattedMessage
      defaultMessage="Personal Settings"
      description="Title of the User Profile personal settings panel"
      id="UserProfile.PersonalSettingsPanel.title"
      values={Object {}}
    />
  }
>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={true}
      title={
        <FormattedMessage
          defaultMessage="Language"
          description="Language label"
          id="UserProfile.PersonalSettingsPanel.language"
          values={Object {}}
        />
      }
    />
    <Stack
      alignItems="stretch"
      scale="s"
    >
      <Select
        addLabelText="Add \\"{label}\\"?"
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        clearAllText="Clear all"
        clearRenderer={[Function]}
        clearValueText="Clear value"
        clearable={false}
        deleteRemoves={true}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        name="language"
        noResultsText="No results found"
        onBlur={[Function]}
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": "English (en)",
              "value": "en",
            },
            Object {
              "label": "German (de)",
              "value": "de",
            },
          ]
        }
        pageSize={5}
        parse={[Function]}
        placeholder="Select..."
        required={false}
        scrollMenuIntoView={true}
        searchable={false}
        simpleValue={false}
        tabSelectsValue={true}
        value="en"
        valueComponent={[Function]}
        valueKey="value"
      />
      <ValidationErrorSwitch
        errors={Object {}}
        isTouched={false}
      >
        <ValidationErrorMatch
          rule="required"
        >
          <ErrorMessage>
            <FormattedMessage
              defaultMessage="Please fill in this required field."
              description="An error message to show if the field is required"
              id="Validation.required"
              values={Object {}}
            />
          </ErrorMessage>
        </ValidationErrorMatch>
      </ValidationErrorSwitch>
    </Stack>
  </FormBox>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={false}
      title={
        <FormattedMessage
          defaultMessage="Time Zone"
          description="TimeZone label"
          id="UserProfile.PersonalSettingsPanel.timeZone"
          values={Object {}}
        />
      }
    />
    <Select
      addLabelText="Add \\"{label}\\"?"
      arrowRenderer={[Function]}
      autosize={true}
      backspaceRemoves={true}
      backspaceToRemoveMessage="Press backspace to remove {label}"
      clearAllText="Clear all"
      clearRenderer={[Function]}
      clearValueText="Clear value"
      clearable={true}
      deleteRemoves={true}
      delimiter=","
      disabled={false}
      escapeClearsValue={true}
      filterOptions={[Function]}
      ignoreAccents={true}
      ignoreCase={true}
      inputProps={Object {}}
      isLoading={false}
      joinValues={false}
      labelKey="label"
      matchPos="any"
      matchProp="any"
      menuBuffer={0}
      menuRenderer={[Function]}
      multi={false}
      name="timeZone"
      noResultsText="No results found"
      onBlur={[Function]}
      onBlurResetsInput={true}
      onChange={[Function]}
      onCloseResetsInput={true}
      optionComponent={[Function]}
      options={
        Array [
          Object {
            "label": "Europe/Berlin - CEST (+02:00)",
            "value": "Europe/Berlin",
          },
        ]
      }
      pageSize={5}
      parse={[Function]}
      placeholder="Select..."
      required={false}
      scrollMenuIntoView={true}
      searchable={true}
      simpleValue={false}
      tabSelectsValue={true}
      value="Europe/Berlin"
      valueComponent={[Function]}
      valueKey="value"
    />
  </FormBox>
</CollapsiblePanel>
`;

exports[`rendering when submitting should disable all input fields 1`] = `
<CollapsiblePanel
  label={
    <FormattedMessage
      defaultMessage="Personal Settings"
      description="Title of the User Profile personal settings panel"
      id="UserProfile.PersonalSettingsPanel.title"
      values={Object {}}
    />
  }
>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={true}
      title={
        <FormattedMessage
          defaultMessage="Language"
          description="Language label"
          id="UserProfile.PersonalSettingsPanel.language"
          values={Object {}}
        />
      }
    />
    <Stack
      alignItems="stretch"
      scale="s"
    >
      <Select
        addLabelText="Add \\"{label}\\"?"
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        clearAllText="Clear all"
        clearRenderer={[Function]}
        clearValueText="Clear value"
        clearable={false}
        deleteRemoves={true}
        delimiter=","
        disabled={true}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        name="language"
        noResultsText="No results found"
        onBlur={[Function]}
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": "English (en)",
              "value": "en",
            },
            Object {
              "label": "German (de)",
              "value": "de",
            },
          ]
        }
        pageSize={5}
        parse={[Function]}
        placeholder="Select..."
        required={false}
        scrollMenuIntoView={true}
        searchable={false}
        simpleValue={false}
        tabSelectsValue={true}
        value="en"
        valueComponent={[Function]}
        valueKey="value"
      />
      <ValidationErrorSwitch
        errors={Object {}}
        isTouched={false}
      >
        <ValidationErrorMatch
          rule="required"
        >
          <ErrorMessage>
            <FormattedMessage
              defaultMessage="Please fill in this required field."
              description="An error message to show if the field is required"
              id="Validation.required"
              values={Object {}}
            />
          </ErrorMessage>
        </ValidationErrorMatch>
      </ValidationErrorSwitch>
    </Stack>
  </FormBox>
  <FormBox
    hasWarning={false}
  >
    <LabelField
      isRequired={false}
      title={
        <FormattedMessage
          defaultMessage="Time Zone"
          description="TimeZone label"
          id="UserProfile.PersonalSettingsPanel.timeZone"
          values={Object {}}
        />
      }
    />
    <Select
      addLabelText="Add \\"{label}\\"?"
      arrowRenderer={[Function]}
      autosize={true}
      backspaceRemoves={true}
      backspaceToRemoveMessage="Press backspace to remove {label}"
      clearAllText="Clear all"
      clearRenderer={[Function]}
      clearValueText="Clear value"
      clearable={true}
      deleteRemoves={true}
      delimiter=","
      disabled={true}
      escapeClearsValue={true}
      filterOptions={[Function]}
      ignoreAccents={true}
      ignoreCase={true}
      inputProps={Object {}}
      isLoading={false}
      joinValues={false}
      labelKey="label"
      matchPos="any"
      matchProp="any"
      menuBuffer={0}
      menuRenderer={[Function]}
      multi={false}
      name="timeZone"
      noResultsText="No results found"
      onBlur={[Function]}
      onBlurResetsInput={true}
      onChange={[Function]}
      onCloseResetsInput={true}
      optionComponent={[Function]}
      options={
        Array [
          Object {
            "label": "Europe/Berlin - CEST (+02:00)",
            "value": "Europe/Berlin",
          },
        ]
      }
      pageSize={5}
      parse={[Function]}
      placeholder="Select..."
      required={false}
      scrollMenuIntoView={true}
      searchable={true}
      simpleValue={false}
      tabSelectsValue={true}
      value="Europe/Berlin"
      valueComponent={[Function]}
      valueKey="value"
    />
  </FormBox>
</CollapsiblePanel>
`;
